File: InterviewPrepare/HouseRobberII.java
/**
 * LeetCode 213 House Robber II
 */
class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        if (nums.length < 4) {
            return Arrays.stream(nums).max().getAsInt();
        }

        int containOne = rangeFind(nums,0,nums.length - 2);
        int containLast = rangeFind(nums,1,nums.length - 1);
        return Math.max(containOne,containLast);
    }

    private int rangeFind(int[] nums, int start,int end) {
        int[] dp = new int[nums.length];
        dp[start] = nums[start];
        dp[start + 1] = Math.max(nums[start], nums[start + 1]);
        for (int i = start + 2; i <= end; i++) {
            dp[i] = Math.max(dp[i - 2] + nums[i],dp[i - 1]);
        }
        return dp[end];
    }
}

File: HouseRobberII.go
func rob(nums []int) int {
	if nums == nil {
		return 0
	}
	n := len(nums)
	if n == 1 {
		return nums[0]
	}
	if n == 2 {
		return max(nums[0],nums[1])
	}
	toRob := func(curs []int) int {
		dp := make([]int,len(curs))
		dp[0],dp[1] = curs[0],max(curs[0],curs[1])
		for i := 2;i < len(curs);i++ {
			dp[i] = max(dp[i - 2] + curs[i],dp[i - 1])
		}
		return dp[len(curs) - 1]
	}
	return max(toRob(nums[:n - 1]),toRob(nums[1:]))
}

func max(a int,b int) int {
	if (a > b) {
		return a
	}
	return b
}
